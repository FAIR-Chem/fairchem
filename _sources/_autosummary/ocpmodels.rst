ocpmodels
=========

.. automodule:: ocpmodels

   
   
   

   
   
   

   
   
   

   
   
   



.. rubric:: Modules

.. autosummary::
   :toctree:
   :recursive:

   ocpmodels.common
   ocpmodels.datasets
   ocpmodels.models
   ocpmodels.modules
   ocpmodels.preprocessing
   ocpmodels.tasks
   ocpmodels.trainers

