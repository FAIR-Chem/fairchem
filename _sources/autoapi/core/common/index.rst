:py:mod:`core.common`
=====================

.. py:module:: core.common

.. autoapi-nested-parse::

   Copyright (c) Meta, Inc. and its affiliates.

   This source code is licensed under the MIT license found in the
   LICENSE file in the root directory of this source tree.



Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   relaxation/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   data_parallel/index.rst
   distutils/index.rst
   flags/index.rst
   gp_utils/index.rst
   hpo_utils/index.rst
   logger/index.rst
   registry/index.rst
   transforms/index.rst
   tutorial_utils/index.rst
   typing/index.rst
   utils/index.rst


