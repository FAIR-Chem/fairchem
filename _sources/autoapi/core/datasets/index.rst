core.datasets
=============

.. py:module:: core.datasets


Subpackages
-----------

.. toctree::
   :maxdepth: 1

   /autoapi/core/datasets/embeddings/index


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/core/datasets/_utils/index
   /autoapi/core/datasets/ase_datasets/index
   /autoapi/core/datasets/base_dataset/index
   /autoapi/core/datasets/lmdb_database/index
   /autoapi/core/datasets/lmdb_dataset/index
   /autoapi/core/datasets/oc22_lmdb_dataset/index
   /autoapi/core/datasets/target_metadata_guesser/index


Classes
-------

.. autoapisummary::

   core.datasets.AseDBDataset
   core.datasets.AseReadDataset
   core.datasets.AseReadMultiStructureDataset
   core.datasets.LMDBDatabase
   core.datasets.LmdbDataset


Functions
---------

.. autoapisummary::

   core.datasets.create_dataset
   core.datasets.data_list_collater


Package Contents
----------------

.. py:class:: AseDBDataset(config: dict, atoms_transform: Callable[[ase.Atoms, Any, Ellipsis], ase.Atoms] = apply_one_tags)

   Bases: :py:obj:`AseAtomsDataset`


   This Dataset connects to an ASE Database, allowing the storage of atoms objects
   with a variety of backends including JSON, SQLite, and database server options.

   For more information, see:
   https://databases.fysik.dtu.dk/ase/ase/db/db.html

   :param config:
                  src (str): Either
                          - the path an ASE DB,
                          - the connection address of an ASE DB,
                          - a folder with multiple ASE DBs,
                          - a list of folders with ASE DBs
                          - a glob string to use to find ASE DBs, or
                          - a list of ASE db paths/addresses.
                          If a folder, every file will be attempted as an ASE DB, and warnings
                          are raised for any files that can't connect cleanly

                          Note that for large datasets, ID loading can be slow and there can be many
                          ids, so it's advised to make loading the id list as easy as possible. There is not
                          an obvious way to get a full list of ids from most ASE dbs besides simply looping
                          through the entire dataset. See the AseLMDBDataset which was written with this usecase
                          in mind.

                  connect_args (dict): Keyword arguments for ase.db.connect()

                  select_args (dict): Keyword arguments for ase.db.select()
                          You can use this to query/filter your database

                  a2g_args (dict): Keyword arguments for fairchem.core.preprocessing.AtomsToGraphs()
                          default options will work for most users

                          If you are using this for a training dataset, set
                          "r_energy":True, "r_forces":True, and/or "r_stress":True as appropriate
                          In that case, energy/forces must be in the database

                  keep_in_memory (bool): Store data in memory. This helps avoid random reads if you need
                          to iterate over a dataset many times (e.g. training for many epochs).
                          Not recommended for large datasets.

                  atoms_transform_args (dict): Additional keyword arguments for the atoms_transform callable

                  transforms (dict[str, dict]): Dictionary specifying data transforms as {transform_function: config}
                          where config is a dictionary specifying arguments to the transform_function

                  key_mapping (dict[str, str]): Dictionary specifying a mapping between the name of a property used
                      in the model with the corresponding property as it was named in the dataset. Only need to use if
                      the name is different.
   :type config: dict
   :param atoms_transform: Additional preprocessing function applied to the Atoms
                           object. Useful for applying tags, for example.
   :type atoms_transform: callable, optional
   :param transform: deprecated?
   :type transform: callable, optional


   .. py:method:: _load_dataset_get_ids(config: dict) -> list[int]


   .. py:method:: get_atoms(idx: int) -> ase.Atoms

      Get atoms object corresponding to datapoint idx. Useful to read other properties not in data object.
      :param idx: index in dataset
      :type idx: int

      :returns: ASE atoms corresponding to datapoint idx
      :rtype: atoms



   .. py:method:: connect_db(address: str | pathlib.Path, connect_args: dict | None = None) -> ase.db.core.Database
      :staticmethod:



   .. py:method:: __del__()


   .. py:method:: sample_property_metadata(num_samples: int = 100) -> dict


   .. py:method:: get_relaxed_energy(identifier)
      :abstractmethod:



.. py:class:: AseReadDataset(config: dict, atoms_transform: Callable[[ase.Atoms, Any, Ellipsis], ase.Atoms] = apply_one_tags)

   Bases: :py:obj:`AseAtomsDataset`


   This Dataset uses ase.io.read to load data from a directory on disk.
   This is intended for small-scale testing and demonstrations of OCP.
   Larger datasets are better served by the efficiency of other dataset types
   such as LMDB.

   For a full list of ASE-readable filetypes, see
   https://wiki.fysik.dtu.dk/ase/ase/io/io.html

   :param config: src (str): The source folder that contains your ASE-readable files

                  pattern (str): Filepath matching each file you want to read
                          ex. "*/POSCAR", "*.cif", "*.xyz"
                          search recursively with two wildcards: "**/POSCAR" or "**/*.cif"

                  a2g_args (dict): Keyword arguments for fairchem.core.preprocessing.AtomsToGraphs()
                          default options will work for most users

                          If you are using this for a training dataset, set
                          "r_energy":True, "r_forces":True, and/or "r_stress":True as appropriate
                          In that case, energy/forces must be in the files you read (ex. OUTCAR)

                  ase_read_args (dict): Keyword arguments for ase.io.read()

                  keep_in_memory (bool): Store data in memory. This helps avoid random reads if you need
                          to iterate over a dataset many times (e.g. training for many epochs).
                          Not recommended for large datasets.

                  include_relaxed_energy (bool): Include the relaxed energy in the resulting data object.
                          The relaxed structure is assumed to be the final structure in the file
                          (e.g. the last frame of a .traj).

                  atoms_transform_args (dict): Additional keyword arguments for the atoms_transform callable

                  transform_args (dict): Additional keyword arguments for the transform callable

                  key_mapping (dict[str, str]): Dictionary specifying a mapping between the name of a property used
                      in the model with the corresponding property as it was named in the dataset. Only need to use if
                      the name is different.
   :type config: dict
   :param atoms_transform: Additional preprocessing function applied to the Atoms
                           object. Useful for applying tags, for example.
   :type atoms_transform: callable, optional


   .. py:method:: _load_dataset_get_ids(config) -> list[pathlib.Path]


   .. py:method:: get_atoms(idx: str | int) -> ase.Atoms


   .. py:method:: get_relaxed_energy(identifier) -> float


.. py:class:: AseReadMultiStructureDataset(config: dict, atoms_transform: Callable[[ase.Atoms, Any, Ellipsis], ase.Atoms] = apply_one_tags)

   Bases: :py:obj:`AseAtomsDataset`


   This Dataset can read multiple structures from each file using ase.io.read.
   The disadvantage is that all files must be read at startup.
   This is a significant cost for large datasets.

   This is intended for small-scale testing and demonstrations of OCP.
   Larger datasets are better served by the efficiency of other dataset types
   such as LMDB.

   For a full list of ASE-readable filetypes, see
   https://wiki.fysik.dtu.dk/ase/ase/io/io.html

   :param config: src (str): The source folder that contains your ASE-readable files

                  pattern (str): Filepath matching each file you want to read
                          ex. "*.traj", "*.xyz"
                          search recursively with two wildcards: "**/POSCAR" or "**/*.cif"

                  index_file (str): Filepath to an indexing file, which contains each filename
                          and the number of structures contained in each file. For instance:

                          /path/to/relaxation1.traj 200
                          /path/to/relaxation2.traj 150

                          This will overrule the src and pattern that you specify!

                  a2g_args (dict): Keyword arguments for fairchem.core.preprocessing.AtomsToGraphs()
                          default options will work for most users

                          If you are using this for a training dataset, set
                          "r_energy":True, "r_forces":True, and/or "r_stress":True as appropriate
                          In that case, energy/forces must be in the files you read (ex. OUTCAR)

                  ase_read_args (dict): Keyword arguments for ase.io.read()

                  keep_in_memory (bool): Store data in memory. This helps avoid random reads if you need
                          to iterate over a dataset many times (e.g. training for many epochs).
                          Not recommended for large datasets.

                  include_relaxed_energy (bool): Include the relaxed energy in the resulting data object.
                          The relaxed structure is assumed to be the final structure in the file
                          (e.g. the last frame of a .traj).

                  use_tqdm (bool): Use TQDM progress bar when initializing dataset

                  atoms_transform_args (dict): Additional keyword arguments for the atoms_transform callable

                  transform_args (dict): Additional keyword arguments for the transform callable

                  key_mapping (dict[str, str]): Dictionary specifying a mapping between the name of a property used
                      in the model with the corresponding property as it was named in the dataset. Only need to use if
                      the name is different.
   :type config: dict
   :param atoms_transform: Additional preprocessing function applied to the Atoms
                           object. Useful for applying tags, for example.
   :type atoms_transform: callable, optional
   :param transform: Additional preprocessing function for the Data object
   :type transform: callable, optional


   .. py:method:: _load_dataset_get_ids(config) -> list[str]


   .. py:method:: get_atoms(idx: str) -> ase.Atoms


   .. py:method:: sample_property_metadata(num_samples: int = 100) -> dict


   .. py:method:: get_relaxed_energy(identifier) -> float


.. py:function:: create_dataset(config: dict[str, Any], split: str) -> Subset

   Create a dataset from a config dictionary

   :param config: dataset config dictionary
   :type config: dict
   :param split: name of split
   :type split: str

   :returns: dataset subset class
   :rtype: Subset


.. py:class:: LMDBDatabase(filename: str | pathlib.Path | None = None, create_indices: bool = True, use_lock_file: bool = False, serial: bool = False, readonly: bool = False, *args, **kwargs)

   Bases: :py:obj:`ase.db.core.Database`


   Base class for all databases.


   .. py:attribute:: readonly


   .. py:attribute:: ids
      :value: []



   .. py:attribute:: deleted_ids
      :value: []



   .. py:method:: __enter__() -> typing_extensions.Self


   .. py:method:: __exit__(exc_type, exc_value, tb) -> None


   .. py:method:: close() -> None


   .. py:method:: _write(atoms: ase.Atoms | ase.db.row.AtomsRow, key_value_pairs: dict, data: dict | None, idx: int | None = None) -> None


   .. py:method:: _update(idx: int, key_value_pairs: dict | None = None, data: dict | None = None)


   .. py:method:: _write_deleted_ids()


   .. py:method:: delete(ids: list[int]) -> None

      Delete rows.



   .. py:method:: _get_row(idx: int, include_data: bool = True)


   .. py:method:: _get_row_by_index(index: int, include_data: bool = True)

      Auxiliary function to get the ith entry, rather than a specific id



   .. py:method:: _select(keys, cmps: list[tuple[str, str, str]], explain: bool = False, verbosity: int = 0, limit: int | None = None, offset: int = 0, sort: str | None = None, include_data: bool = True, columns: str = 'all')


   .. py:property:: metadata
      Load the metadata from the DB if present


   .. py:property:: _nextid
      Get the id of the next row to be written


   .. py:method:: count(selection=None, **kwargs) -> int

      Count rows.

      See the select() method for the selection syntax.  Use db.count() or
      len(db) to count all rows.



   .. py:method:: _load_ids() -> None

      Load ids from the DB

      Since ASE db ids are mostly 1-N integers, but can be missing entries
      if ids have been deleted. To save space and operating under the assumption
      that there will probably not be many deletions in most OCP datasets,
      we just store the deleted ids.



.. py:class:: LmdbDataset(config)

   Bases: :py:obj:`fairchem.core.datasets.base_dataset.BaseDataset`


   Base Dataset class for all OCP datasets.


   .. py:attribute:: sharded
      :type:  bool

      Dataset class to load from LMDB files containing relaxation
      trajectories or single point computations.
      Useful for Structure to Energy & Force (S2EF), Initial State to
      Relaxed State (IS2RS), and Initial State to Relaxed Energy (IS2RE) tasks.
      The keys in the LMDB must be integers (stored as ascii objects) starting
      from 0 through the length of the LMDB. For historical reasons any key named
      "length" is ignored since that was used to infer length of many lmdbs in the same
      folder, but lmdb lengths are now calculated directly from the number of keys.
      :param config: Dataset configuration
      :type config: dict


   .. py:attribute:: path


   .. py:attribute:: key_mapping


   .. py:attribute:: transforms


   .. py:method:: __getitem__(idx: int) -> T_co


   .. py:method:: connect_db(lmdb_path: pathlib.Path | None = None) -> lmdb.Environment


   .. py:method:: __del__()


   .. py:method:: sample_property_metadata(num_samples: int = 100)


.. py:function:: data_list_collater(data_list: list[torch_geometric.data.data.BaseData], otf_graph: bool = False, to_dict: bool = False) -> torch_geometric.data.data.BaseData | dict[str, torch.Tensor]

