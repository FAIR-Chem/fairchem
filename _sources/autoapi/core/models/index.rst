:py:mod:`core.models`
=====================

.. py:module:: core.models


Subpackages
-----------
.. toctree::
   :titlesonly:
   :maxdepth: 3

   equiformer_v2/index.rst
   escn/index.rst
   gemnet/index.rst
   gemnet_gp/index.rst
   gemnet_oc/index.rst
   painn/index.rst
   scn/index.rst
   utils/index.rst


Submodules
----------
.. toctree::
   :titlesonly:
   :maxdepth: 1

   base/index.rst
   dimenet_plus_plus/index.rst
   model_registry/index.rst
   schnet/index.rst


Package Contents
----------------


Functions
~~~~~~~~~

.. autoapisummary::

   core.models.model_name_to_local_file



Attributes
~~~~~~~~~~

.. autoapisummary::

   core.models.available_pretrained_models


.. py:data:: available_pretrained_models

   

.. py:function:: model_name_to_local_file(model_name: str, local_cache: str | pathlib.Path) -> str

   Download a pretrained checkpoint if it does not exist already

   :param model_name: the model name. See available_pretrained_checkpoints.
   :type model_name: str
   :param local_cache: path to local cache directory
   :type local_cache: str or Path

   :returns: local path to checkpoint file
   :rtype: str


