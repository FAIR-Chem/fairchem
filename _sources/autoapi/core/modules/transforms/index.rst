:py:mod:`core.modules.transforms`
=================================

.. py:module:: core.modules.transforms


Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   core.modules.transforms.DataTransforms



Functions
~~~~~~~~~

.. autoapisummary::

   core.modules.transforms.decompose_tensor



.. py:class:: DataTransforms(config)


   .. py:method:: __call__(data_object)



.. py:function:: decompose_tensor(data_object, config) -> torch_geometric.data.Data


