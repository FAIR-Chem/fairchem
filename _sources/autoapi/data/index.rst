data
====

.. py:module:: data


Subpackages
-----------

.. toctree::
   :maxdepth: 1

   /autoapi/data/oc/index
   /autoapi/data/odac/index
   /autoapi/data/om/index


