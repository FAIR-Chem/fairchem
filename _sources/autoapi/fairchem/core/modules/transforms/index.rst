:py:mod:`fairchem.core.modules.transforms`
==========================================

.. py:module:: fairchem.core.modules.transforms


Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   fairchem.core.modules.transforms.DataTransforms



Functions
~~~~~~~~~

.. autoapisummary::

   fairchem.core.modules.transforms.decompose_tensor



.. py:class:: DataTransforms(config)


   .. py:method:: __call__(data_object)



.. py:function:: decompose_tensor(data_object, config) -> torch_geometric.data.Data


