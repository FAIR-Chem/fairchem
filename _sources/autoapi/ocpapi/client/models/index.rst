:py:mod:`ocpapi.client.models`
==============================

.. py:module:: ocpapi.client.models


Module Contents
---------------

Classes
~~~~~~~

.. autoapisummary::

   ocpapi.client.models._DataModel
   ocpapi.client.models.Model
   ocpapi.client.models.Models
   ocpapi.client.models.Bulk
   ocpapi.client.models.Bulks
   ocpapi.client.models.Adsorbates
   ocpapi.client.models.Atoms
   ocpapi.client.models.SlabMetadata
   ocpapi.client.models.Slab
   ocpapi.client.models.Slabs
   ocpapi.client.models.AdsorbateSlabConfigs
   ocpapi.client.models.AdsorbateSlabRelaxationsSystem
   ocpapi.client.models.AdsorbateSlabRelaxationsRequest
   ocpapi.client.models.Status
   ocpapi.client.models.AdsorbateSlabRelaxationResult
   ocpapi.client.models.AdsorbateSlabRelaxationsResults




.. py:class:: _DataModel


   Base class for all data models.

   .. py:attribute:: other_fields
      :type: dataclasses_json.CatchAll

      Fields that may have been added to the API that all not yet supported
      explicitly in this class.


.. py:class:: Model


   Bases: :py:obj:`_DataModel`

   Stores information about a single model supported in the API.

   .. py:attribute:: id
      :type: str

      The ID of the model.


.. py:class:: Models


   Bases: :py:obj:`_DataModel`

   Stores the response from a request for models supported in the API.

   .. py:attribute:: models
      :type: List[Model]

      The list of models that are supported.


.. py:class:: Bulk


   Bases: :py:obj:`_DataModel`

   Stores information about a single bulk material.

   .. py:attribute:: src_id
      :type: str

      The ID of the material.

   .. py:attribute:: formula
      :type: str

      The chemical formula of the material.

   .. py:attribute:: elements
      :type: List[str]

      The list of elements in the material.


.. py:class:: Bulks


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to fetch bulks supported in the API.

   .. py:attribute:: bulks_supported
      :type: List[Bulk]

      List of bulks that can be used in the API.


.. py:class:: Adsorbates


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to fetch adsorbates supported in the
   API.

   .. py:attribute:: adsorbates_supported
      :type: List[str]

      List of adsorbates that can be used in the API.


.. py:class:: Atoms


   Bases: :py:obj:`_DataModel`

   Subset of the fields from an ASE Atoms object that are used within this
   API.

   .. py:attribute:: cell
      :type: Tuple[Tuple[float, float, float], Tuple[float, float, float], Tuple[float, float, float]]

      3x3 matrix with unit cell vectors.

   .. py:attribute:: pbc
      :type: Tuple[bool, bool, bool]

      Whether the structure is periodic along the a, b, and c lattice vectors,
      respectively.

   .. py:attribute:: numbers
      :type: List[int]

      The atomic number of each atom in the unit cell.

   .. py:attribute:: positions
      :type: List[Tuple[float, float, float]]

      The coordinates of each atom in the unit cell, relative to the cartesian
      frame.

   .. py:attribute:: tags
      :type: List[int]

      Labels for each atom in the unit cell where 0 represents a subsurface atom
      (fixed during optimization), 1 represents a surface atom, and 2 represents
      an adsorbate atom.

   .. py:method:: to_ase_atoms() -> ase.Atoms

      Creates an ase.Atoms object with the positions, element numbers,
      etc. populated from values on this object.

      :returns: ase.Atoms object with values from this object.



.. py:class:: SlabMetadata


   Bases: :py:obj:`_DataModel`

   Stores metadata about a slab that is returned from the API.

   .. py:attribute:: bulk_src_id
      :type: str

      The ID of the bulk material from which the slab was derived.

   .. py:attribute:: millers
      :type: Tuple[int, int, int]

      The Miller indices of the slab relative to bulk structure.

   .. py:attribute:: shift
      :type: float

      The position along the vector defined by the Miller indices at which a
      cut was taken to generate the slab surface.

   .. py:attribute:: top
      :type: bool

      If False, the top and bottom surfaces for this millers/shift pair are
      distinct and this slab represents the bottom surface.


.. py:class:: Slab


   Bases: :py:obj:`_DataModel`

   Stores all information about a slab that is returned from the API.

   .. py:attribute:: atoms
      :type: Atoms

      The structure of the slab.

   .. py:attribute:: metadata
      :type: SlabMetadata

      Extra information about the slab.


.. py:class:: Slabs


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to fetch slabs for a bulk structure.

   .. py:attribute:: slabs
      :type: List[Slab]

      The list of slabs that were generated from the input bulk structure.


.. py:class:: AdsorbateSlabConfigs


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to fetch placements of a single
   absorbate on a slab.

   .. py:attribute:: adsorbate_configs
      :type: List[Atoms]

      List of structures, each representing one possible adsorbate placement.

   .. py:attribute:: slab
      :type: Slab

      The structure of the slab on which the adsorbate is placed.


.. py:class:: AdsorbateSlabRelaxationsSystem


   Bases: :py:obj:`_DataModel`

   Stores the response from a request to submit a new batch of adsorbate
   slab relaxations.

   .. py:attribute:: system_id
      :type: str

      Unique ID for this set of relaxations which can be used to fetch results
      later.

   .. py:attribute:: config_ids
      :type: List[int]

      The list of IDs assigned to each of the input adsorbate placements, in the
      same order in which they were submitted.


.. py:class:: AdsorbateSlabRelaxationsRequest


   Bases: :py:obj:`_DataModel`

   Stores the request to submit a new batch of adsorbate slab relaxations.

   .. py:attribute:: adsorbate
      :type: str

      Description of the adsorbate.

   .. py:attribute:: adsorbate_configs
      :type: List[Atoms]

      List of adsorbate placements being relaxed.

   .. py:attribute:: bulk
      :type: Bulk

      Information about the original bulk structure used to create the slab.

   .. py:attribute:: slab
      :type: Slab

      The structure of the slab on which adsorbates are placed.

   .. py:attribute:: model
      :type: str

      The type of the ML model being used during relaxations.

   .. py:attribute:: ephemeral
      :type: Optional[bool]

      Whether the relaxations can be deleted (assume they cannot be deleted if
      None).

   .. py:attribute:: adsorbate_reaction
      :type: Optional[str]

      If possible, an html-formatted string describing the reaction will be added
      to this field.


.. py:class:: Status(*args, **kwds)


   Bases: :py:obj:`enum.Enum`

   Relaxation status of a single adsorbate placement on a slab.

   .. py:attribute:: NOT_AVAILABLE
      :value: 'not_available'

      The configuration exists but the result is not yet available. It is
      possible that checking again in the future could yield a result.

   .. py:attribute:: FAILED_RELAXATION
      :value: 'failed_relaxation'

      The relaxation failed for this configuration.

   .. py:attribute:: SUCCESS
      :value: 'success'

      The relaxation was successful and the requested information about the
      configuration was returned.

   .. py:attribute:: DOES_NOT_EXIST
      :value: 'does_not_exist'

      The requested configuration does not exist.

   .. py:method:: __str__() -> str

      Return str(self).



.. py:class:: AdsorbateSlabRelaxationResult


   Bases: :py:obj:`_DataModel`

   Stores information about a single adsorbate slab configuration, including
   outputs for the model used in relaxations.

   The API to fetch relaxation results supports requesting a subset of fields
   in order to limit the size of response payloads. Optional attributes will
   be defined only if they are including the response.

   .. py:attribute:: config_id
      :type: int

      ID of the configuration within the system.

   .. py:attribute:: status
      :type: Status

      The status of the request for information about this configuration.

   .. py:attribute:: system_id
      :type: Optional[str]

      The ID of the system in which the configuration was originally submitted.

   .. py:attribute:: cell
      :type: Optional[Tuple[Tuple[float, float, float], Tuple[float, float, float], Tuple[float, float, float]]]

      3x3 matrix with unit cell vectors.

   .. py:attribute:: pbc
      :type: Optional[Tuple[bool, bool, bool]]

      Whether the structure is periodic along the a, b, and c lattice vectors,
      respectively.

   .. py:attribute:: numbers
      :type: Optional[List[int]]

      The atomic number of each atom in the unit cell.

   .. py:attribute:: positions
      :type: Optional[List[Tuple[float, float, float]]]

      The coordinates of each atom in the unit cell, relative to the cartesian
      frame.

   .. py:attribute:: tags
      :type: Optional[List[int]]

      Labels for each atom in the unit cell where 0 represents a subsurface atom
      (fixed during optimization), 1 represents a surface atom, and 2 represents
      an adsorbate atom.

   .. py:attribute:: energy
      :type: Optional[float]

      The energy of the configuration.

   .. py:attribute:: energy_trajectory
      :type: Optional[List[float]]

      The energy of the configuration at each point along the relaxation
      trajectory.

   .. py:attribute:: forces
      :type: Optional[List[Tuple[float, float, float]]]

      The forces on each atom in the relaxed structure.

   .. py:method:: to_ase_atoms() -> ase.Atoms

      Creates an ase.Atoms object with the positions, element numbers,
      etc. populated from values on this object.

      The predicted energy and forces will also be copied to the new
      ase.Atoms object as a SinglePointCalculator (a calculator that
      stores the results of an already-run simulation).

      :returns: ase.Atoms object with values from this object.



.. py:class:: AdsorbateSlabRelaxationsResults


   Bases: :py:obj:`_DataModel`

   Stores the response from a request for results of adsorbate slab
   relaxations.

   .. py:attribute:: configs
      :type: List[AdsorbateSlabRelaxationResult]

      List of configurations in the system, each representing one placement of
      an adsorbate on a slab surface.

   .. py:attribute:: omitted_config_ids
      :type: List[int]

      List of IDs of configurations that were requested but omitted by the
      server. Results for these IDs can be requested again.


