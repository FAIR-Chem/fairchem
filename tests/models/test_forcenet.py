"""
Copyright (c) Facebook, Inc. and its affiliates.

This source code is licensed under the MIT license found in the
LICENSE file in the root directory of this source tree.
"""

import os
import random

import numpy as np
import pytest
import torch
from ase.io import read
from torch_geometric.data import Data

from ocpmodels.common.transforms import RandomRotate
from ocpmodels.datasets import data_list_collater
from ocpmodels.models import ForceNet
from ocpmodels.preprocessing import AtomsToGraphs


@pytest.fixture(scope="class")
def load_data(request):
    atoms = read(
        os.path.join(os.path.dirname(os.path.abspath(__file__)), "atoms.json"),
        index=0,
        format="json",
    )
    a2g = AtomsToGraphs(
        max_neigh=200,
        radius=6,
        r_energy=True,
        r_forces=True,
        r_distances=True,
    )
    data_list = a2g.convert_all([atoms])
    request.cls.data = data_list[0]


@pytest.fixture(scope="class")
def load_model(request):
    model = ForceNet(
        None,
        32,
        1,
        cutoff=6.0,
    )
    request.cls.model = model


@pytest.mark.usefixtures("load_data")
@pytest.mark.usefixtures("load_model")
class TestForceNet:
    def test_energy_force_shape(self):
        data = self.data

        # Pass it through the model.
        out = self.model(data_list_collater([data]))

        # Compare shape of predicted energies, forces.
        energy = out[0].detach()
        np.testing.assert_equal(energy.shape, (1, 1))

        forces = out[1].detach()
        np.testing.assert_equal(forces.shape, (data.pos.shape[0], 3))
