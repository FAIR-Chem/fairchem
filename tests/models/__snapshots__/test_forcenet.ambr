# name: TestForceNet.test_energy_force_shape
  Size(
    1,
    1,
  )
# ---
# name: TestForceNet.test_energy_force_shape.1
  Approx(
  	array([[5.4464083]], dtype=float32)
  )
# ---
# name: TestForceNet.test_energy_force_shape.2
  Size(
    34,
    3,
  )
# ---
# name: TestForceNet.test_energy_force_shape.3
  Approx(
  	array([[-1.30544817e+00, -1.49980092e+00,  2.58076763e+00],
  	       [-1.56487048e-01,  1.21984780e-01,  5.30002713e-01],
  	       [ 2.38139129e+00,  4.72385049e-01,  2.54773617e+00],
  	       [-4.32623863e-01, -4.87135053e-01,  2.76770496e+00],
  	       [ 1.39428461e+00,  6.48864388e-01,  6.26993179e-01],
  	       [ 4.07347471e-01,  1.23799689e-01, -5.07026792e-01],
  	       [ 2.25082827e+00,  9.96077657e-02,  1.82430422e+00],
  	       [-3.89873087e-01, -5.24788022e-01,  2.83203650e+00],
  	       [-7.49948323e-02,  1.06079209e+00,  5.90661526e-01],
  	       [ 3.96642715e-01,  1.45493388e-01, -4.97148573e-01],
  	       [-2.75852084e-02, -1.35263979e+00,  3.46320987e-01],
  	       [ 1.14955910e-01,  1.54354393e-01,  7.60127604e-01],
  	       [ 9.49162841e-01, -9.56393898e-01, -1.52243674e-01],
  	       [ 6.18018210e-01, -8.03342938e-01,  4.35517371e-01],
  	       [ 1.03898096e+00, -7.93570936e-01, -1.10788167e-01],
  	       [-9.31364298e-02, -1.40183270e+00,  2.39522368e-01],
  	       [ 6.57854915e-01, -4.57121521e-01,  3.02878112e-01],
  	       [ 1.14108153e-01,  1.36927634e-01,  7.50466406e-01],
  	       [ 5.74506104e-01, -2.99378604e-01, -1.67183876e-01],
  	       [ 1.42032638e-01, -1.20734669e-01,  7.83950210e-01],
  	       [-1.45273298e-01,  1.05403699e-01,  1.37604153e+00],
  	       [ 2.53411889e-01, -2.43154675e-01,  3.51213545e-01],
  	       [ 5.22605896e-01, -3.22370619e-01, -1.23302817e-01],
  	       [ 1.19088799e-01, -1.06448963e-01,  7.81443238e-01],
  	       [-4.69261706e-02, -2.09370628e-03,  1.46875441e+00],
  	       [ 2.96553016e-01, -3.83107215e-01,  5.76262951e-01],
  	       [ 6.58208549e-01, -1.03099659e-01, -1.22475326e-01],
  	       [ 1.00546025e-01, -8.82540047e-02,  9.23259616e-01],
  	       [-1.48815066e-01, -2.76506990e-01,  1.26217115e+00],
  	       [ 3.05881441e-01, -2.15618253e-01,  2.10024521e-01],
  	       [ 1.36593431e-01, -3.22298646e-01,  1.27883935e+00],
  	       [ 6.13107383e-01, -5.51185012e-02, -2.86964476e-02],
  	       [ 2.06872270e-01, -1.47157997e-01,  2.15230256e-01],
  	       [ 7.88176805e-02, -8.68521333e-02,  9.16256189e-01]], dtype=float32)
  )
# ---
